### PR TITLE
feat: add slotted poster image for faster load/SSR

### PR DESCRIPTION
**Pull request by luwes**
https://recordit.co/WIRdGQ3bZv

- [x] add poster image thumbhash
- [x] create responsive img[srcset]

### CODE REVIEW

File: src/handlers/local-upload.ts [hunk 1]
```diff
@@ -60,6 +61,14 @@ async function pollForAssetReady(filePath: string, asset: Asset) {
   }
 
   if (muxAsset.status === 'ready') {
+
+    let blurDataURL;
+    try {
+      blurDataURL = await createThumbHash(`https://image.mux.com/${playbackId}/thumbnail.png?width=100&height=100`);
```
Comment (mmcc) [inline src/handlers/local-upload.ts:67]: Does it matter that this is going to distort the aspect ratio? Should we just specify height or width to maintain it?

File: src/handlers/local-upload.ts [hunk 2]
```diff
@@ -68,6 +77,7 @@ async function pollForAssetReady(filePath: string, asset: Asset) {
       externalIds: {
         playbackId,
       },
+      blurDataURL,
```
Comment (mmcc) [inline src/handlers/local-upload.ts:80]: Feels a little weird to put this at the top level? Should we consider making `externalIds` something like `uploaderMetadata` or something like that and putting this there too?

File: src/handlers/local-upload.ts [hunk 3]
```diff
@@ -178,3 +188,20 @@ export default async function uploadLocalFile(asset: Asset) {
...
+    .ensureAlpha()
+    .toBuffer({ resolveWithObject: true });
+
+  // thumbhash is ESM only so dynamically import it.
+  const { rgbaToThumbHash, thumbHashToDataURL } = await import('thumbhash');
+
+  const hash = rgbaToThumbHash(info.width, info.height, data);
+
+  return thumbHashToDataURL(hash);
+}
```
Comment (mmcc) [inline src/handlers/local-upload.ts:207]: cooool

File: src/next-video.tsx [hunk 4]
```diff
@@ -153,3 +164,55 @@ function Alert({ status, hidden }: AlertProps) {
...
+  const jsonPayload = decodeURIComponent(
+    atob(base64)
+      .split('')
+      .map(function (c) {
+        return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
+      })
+      .join('')
+  );
+  return JSON.parse(jsonPayload);
+}
```
Comment (mmcc) [inline src/next-video.tsx:218]: Does this add signed playback URLs?

File: tsconfig.json [hunk 5]
```diff
@@ -10,7 +10,7 @@
     "outDir": "dist",
     "esModuleInterop": true,
     "module": "esnext",
-    "moduleResolution": "nodenext",
+    "moduleResolution": "node",
```
Comment (mmcc) [inline tsconfig.json:13]: I'm not going to lie, I have no idea what this means anymore

**Review by mmcc — APPROVED** [review]
:fire:

File: src/handlers/local-upload.ts (hunk 1 above)
Comment (luwes) [reply src/handlers/local-upload.ts:67]: no, they're max dimensions. it keeps aspect ratio 
> By default, Mux will preserve the aspect ratio of the video

Comment (mmcc) [reply src/handlers/local-upload.ts:67]: Let's pretend I didn't ask that.

File: src/handlers/local-upload.ts (hunk 2 above)
Comment (luwes) [reply src/handlers/local-upload.ts:80]: I agree, that sounds good to me

File: src/next-video.tsx (hunk 4 above)
Comment (luwes) [reply src/next-video.tsx:218]: it parses the JWT and then we can check if it is for sure a thumbnail `aud !== 't'`.
we do it in Mux player as well.

File: tsconfig.json (hunk 5 above)
Comment (luwes) [reply tsconfig.json:13]: I had to change this due to this TS error `JSX element type 'MuxPlayer' does not have any construct or call signatures.`. `nodenext` should be used nowadays according to TS docs but we might doing something wrong in Mux elements, not 100% sure.

https://www.typescriptlang.org/tsconfig#moduleResolution

**Pull request merged**